feat: add ideamarket plugin

MountVase opened the pull request:
### Plugin to display user Ideamarket ranking & token price

fix https://github.com/Ideamarket/gitcoin-bounties/issues/19

#### Screenshots:
![ideamarket_preview_light](https://user-images.githubusercontent.com/63966146/112465792-c8845700-8d6d-11eb-881d-99d060d338cb.png)
![ideamarket_listed_light](https://user-images.githubusercontent.com/63966146/112465808-cd490b00-8d6d-11eb-898f-0aa70b85bc6b.png)
![ideamarket_listed_dark](https://user-images.githubusercontent.com/63966146/112465819-cfab6500-8d6d-11eb-9140-79b23aeef790.png)
![ideamarket_unlisted_light](https://user-images.githubusercontent.com/63966146/112465827-d2a65580-8d6d-11eb-860f-b83aaabf4ba8.png)

MountVase committed:
add ideamarket plugin

MountVase commented:
Also, button placement is something I'm currently still working on (struggling on) .

yisiliu commented:
@MountVase awesome job! Again I think it's worth letting @xiazhiming take a look if he's got some ideas on the UI.

yisiliu commented:
> Also, button placement is something I'm currently still working on (struggling on) .

and does @Jack-Works have any ideas to help @MountVase ?

Tedko commented:
Good jobs.
Guys one small question is how can we make the triggering event not so
‘annoying’? Guess you know what I mean 🤔

On Thu, Mar 25, 2021 at 20:26 Yisi Liu ***@***.***> wrote:

> @MountVase <https://github.com/MountVase> awesome job! Again I think it's
> worth letting @xiazhiming <https://github.com/xiazhiming> take a look if
> he's got some ideas on the UI.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DimensionDev/Maskbook/pull/2665#issuecomment-806645145>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTAVTID3TI2DTLXYGEC7Y3TFMTX3ANCNFSM4ZZFMM7Q>
> .
>

MountVase commented:
> Good jobs. Guys one small question is how can we make the triggering event not so ‘annoying’? Guess you know what I mean 
> […](#)
> On Thu, Mar 25, 2021 at 20:26 Yisi Liu ***@***.***> wrote: @MountVase <https://github.com/MountVase> awesome job! Again I think it's worth letting @xiazhiming <https://github.com/xiazhiming> take a look if he's got some ideas on the UI. — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <[#2665 (comment)](https://github.com/DimensionDev/Maskbook/pull/2665#issuecomment-806645145)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABTAVTID3TI2DTLXYGEC7Y3TFMTX3ANCNFSM4ZZFMM7Q> .

Yeah, I know what you mean :) I'll try to extend the hover area

Tedko commented:
> > Good jobs. Guys one small question is how can we make the triggering event not so ‘annoying’? Guess you know what I mean
> 
> Yeah, I know what you mean :) I'll try to extend the hover area

Thx! good job guys

MountVase committed:
fix hover action, add svg

MountVase commented:
Improved the hovering action a lot in 46e0ce4c3f8419e642f05dfed963d853240a973f. Also, added the real svg from Ideamarket, but couldn't figure out how to load svgs. Added the type in env.d.ts and a line to load svgs with 'file-loader' in webpack config. 

Still doesn't show up though..? But you guys know your webpackconfig better than I do :)

septs commented on packages/maskbook/src/env.d.ts:
```diff
@@ -125,3 +125,8 @@ declare module '*.jpg' {
 }
+
+declare module '*.svg' {
+    const content: any
```
@Jack-Works we allow `any`?

MountVase replied:
It can be string also, loading the actual svg still doesn't work currently.

septs replied:
```ts
new URL('path', import.meta.url)
```

septs replied:
https://github.com/DimensionDev/Maskbook/blob/643a95b9d3e3053a41cb5392008287fe5173c894/packages/maskbook/src/resources/image-payload/index.ts#L2-L10

septs commented on packages/maskbook/webpack.config.ts:
```diff
@@ -111,6 +111,8 @@ function config(opts: {
                 // Opt in source map
                 { test: /(async-call|webextension).+\.js$/, enforce: 'pre', use: ['source-map-loader'] },
+                // support for svg
+                 { test: /\.svg$/, loader: 'file-loader' }, 
```
CC @Jack-Works

Jack-Works replied:
Hi please remove this

MountVase committed:
major changes. Rank displayed, varying icons, new fetching

MountVase commented:
Changed quite a lot of the plugin in fa16974d5b2ad3fa343af2f049bc3ab961ffa855. Instead of fetching each users ranking on each tweet, we fetch all the ideaTokens on render and then query from that data. This makes it possible to display rank & varying icons before even interacting with the icon :)

Tried to keep application rendering logic relatively similar to previous commits. Didn't get the Ideamarket SVGs working still, but demonstrated the functionality with other icons that I got working.
screenshots: 

![ideamarkets snowden](https://user-images.githubusercontent.com/63966146/112719807-7e34de80-8f03-11eb-9087-cef94ff89583.png)
![ideamarkets duncan](https://user-images.githubusercontent.com/63966146/112719809-812fcf00-8f03-11eb-960f-4cc56233bd51.png)

Jack-Works commented on packages/maskbook/src/env.d.ts:
```diff
@@ -125,3 +125,8 @@ declare module '*.jpg' {
     export default content
 }
+
+declare module '*.svg' {
```
Hello please remove those. We're using `new URL(..., import.meta.url)` to load resources.

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/define.tsx:
```diff
@@ -0,0 +1,58 @@
+let ideaTokens: getAllListingsData[]
+
+const fetchAll = async () => {
+    const result = await getAllListings()
```
What if the dataset is very large?

MountVase replied:
Then we'd probably be better off with my previous solution..

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/define.tsx:
```diff
@@ -0,0 +1,58 @@
+    scope: PluginScope.Internal,
+
+    postInspector: function Component(): JSX.Element | null {
+        const user = usePostInfoDetails('postBy').toText()
```
```suggestion
        const user = usePostInfoDetails('postBy').userId
```

Why do you need to parse this object by yourself?

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/Listed.tsx:
```diff
@@ -0,0 +1,127 @@
+    return (
+        <div
+            className={classes.root}
+            onMouseEnter={() => props.setExtendedHover(true)}
```
Can you use [Tooltip](https://next.material-ui.com/components/tooltips/#main-content) instead?

MountVase replied:
Didn't know about that component, should probably reduce some lines of code. I Will not rewrite everything right now to use ToolTip though, as it works.

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/Listed.tsx:
```diff
@@ -0,0 +1,127 @@
+            </div>
+
+            <Divider />
+            <Box display="flex" bgcolor="#f7f7f7" justifyContent="center">
```
```suggestion
            <Box display="flex" sx={{ backgroundColor: "#f7f7f7" }} justifyContent="center">
```

`bgcolor` is deprecated HTML property

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/Listing.tsx:
```diff
@@ -0,0 +1,30 @@
+
+    const rank = Number(props.rank)
+    const price = Number(props.price).toFixed(2)
+    const dayChange = Number(Number(props.dayChange).toFixed(2))
```
Why not
```suggestion
    const dayChange = Number(props.dayChange).toFixed(2)
```

MountVase replied:
.toFixed returns a string

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/Loading.tsx:
```diff
@@ -0,0 +1,53 @@
+                <CircularProgress />
+            </Box>
+
+            <Box display="flex" bgcolor="#f7f7f7" justifyContent="center">
```
bgcolor is deprecated HTML property

MountVase committed:
remove webpack config

MountVase committed:
review fixes

MountVase committed:
prettier fix

ManyRios commented:
Very good job dude!

septs commented on packages/maskbook/src/plugins/Ideamarket/constants.ts:
```diff
@@ -0,0 +1 @@
+export const SUBGRAPH_URI = 'https://subgraph.backend.ideamarket.io:443/subgraphs/name/Ideamarket/Ideamarket'
```
```suggestion
// proxy: https://subgraph.backend.ideamarket.io/subgraphs/*
export const SUBGRAPH_URI = 'https://ideamarket.r2d2.to/subgraphs/name/Ideamarket/Ideamarket'
```

septs replied:
CC @yisiliu plz update dns record.

septs replied:
@MountVase

septs commented on packages/maskbook/src/plugins/Ideamarket/README.md:
```diff
@@ -0,0 +1,27 @@
+
+[graphql-request](https://github.com/prisma-labs/graphql-request) is used instead, which works well but without the caching abilities.
+
+### Issues
```
Remove the section

septs commented on packages/maskbook/src/plugins/Ideamarket/README.md:
```diff
@@ -0,0 +1,27 @@
+
+### Components
+
+Rendering logic (inside /UI directory): Each tweet has a <LogoButton> that's passed a username prop. That component renders a <Listing> when hovered on or clicked. Then <Listing> queries the graphQL endpoint, and returns <Loading> if loading, <Listed> if listing is found, or <NotListed> if it isn't.
```
```suggestion
Rendering logic (inside /UI directory):
Each tweet has a `<LogoButton>` that's passed a username prop.
That component renders a `<Listing>` when hovered on or clicked.
Then `<Listing>` queries the graphQL endpoint,
and returns `<Loading>`
if loading, `<Listed>`
if listing is found, or `<NotListed>` if it isn't.
```

septs commented on packages/maskbook/src/plugins/Ideamarket/README.md:
```diff
@@ -0,0 +1,27 @@
+# Valuables Plugin
+
+Ideamarket.io <https://ideamarket.io/>
```
```suggestion
Ideamarket.io <https://ideamarket.io>
```

MountVase committed:
fetching change, icons fixed.

MountVase committed:
review fixes & improvements

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/types.ts:
```diff
@@ -0,0 +1,22 @@
+    price: string
+}
+
+export interface getAllListingsData {
```
I tend to use Uppercase name for interfaces and types

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/Loading.tsx:
```diff
@@ -0,0 +1,53 @@
+                <CircularProgress />
+            </Box>
+
+            <Box display="flex" sx={{ backgroundColor: '#f7f7f7' }} justifyContent="center">
```
are they work in both light and dark themes?

MountVase replied:
Yes. There was no explicit dark theme design, and I think this one looks good for now.

Tedko committed:
Update packages/maskbook/src/plugins/Ideamarket/README.md

Co-authored-by: Septs <github@septs.pw>

Jack-Works committed:
Merge branch 'master' into pr/MountVase/2665

Jack-Works committed:
chore: remove unused import

Jack-Works commented:
@MountVase hi you can use `useCurrentIdentity` to access the current user id

Tedko commented:
@MountVase hi are you still there?

guanbinrui commented on packages/maskbook/src/plugins/Ideamarket/UI/Loading.tsx:
```diff
@@ -0,0 +1,53 @@
+
+            <Box display="flex" sx={{ backgroundColor: '#f7f7f7' }} justifyContent="center">
+                <Typography className={classes.bottomText}>
+                    powered by <b>Ideamarket</b>
```
Add a hyper link for this tag?

MountVase replied:
Good idea, no need though. Ideamarket guys asked the 'powered by' to be removed, and the Loading component isn't really rendered anymore too.

guanbinrui commented on packages/maskbook/src/plugins/Ideamarket/UI/NotListed.tsx:
```diff
@@ -0,0 +1,64 @@
+        color: 'white',
+    },
+    bottomText: {
+        fontSize: [12, '!important'],
```
COOL!

MountVase replied:
:)

guanbinrui commented on packages/maskbook/src/plugins/Ideamarket/UI/assets/idea-markets-logo.svg:
```diff
@@ -0,0 +1 @@
+<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="473.69" height="700" viewBox="0 0 473.69 700"><defs><style>.cls-1{fill:url(#linear-gradient);}</style><linearGradient id="linear-gradient" x1="306.46" y1="236.5" x2="754.9" y2="725.79" gradientUnits="userSpaceOnUse"><stop offset="0" stop-color="#0e8fff"/><stop offset="0.19" s ...
```
Format it.

MountVase replied:
File wasn't used anymore, svg:s are inside assets/index.tsx, so I deleted it.

guanbinrui commented on packages/maskbook/src/plugins/Ideamarket/api.ts:
```diff
@@ -0,0 +1,51 @@
+        ideaTokens(where: { name: "${name}" }) {
+          rank
+          dayChange
+	  latestPricePoint {
```
Format it.

MountVase commented:
> @MountVase hi are you still there?

Hey, yes I am. I'll take a look at the reviews.

MountVase committed:
review fixes.

MountVase committed:
Merge branch 'plugins/ideamarket' of https://github.com/MountVase/Maskbook into plugins/ideamarket

Bobface committed:
chore: merge upstream master

pbteja1998 committed:
chore: fix the positioning of the Ideamarket button and overlay

pbteja1998 committed:
chore: change colors and add zIndex

pbteja1998 committed:
chore: add 0.3s delay between hover and popup

pbteja1998 committed:
chore: change the divider colors

pbteja1998 committed:
chore: remove rank from overlay

pbteja1998 committed:
chore: change the hover color of buy and list buttons

Bobface committed:
feat: add z-index to not listed popup

Bobface committed:
feat: add seperate onclick handler

Bobface committed:
feat: lighter border color

MountVase committed:
Merge pull request #3 from Ideamarket/ideamarket-bugfix

Ideamarket plugin bugfix

Bobface commented:
Hey all, the plugin is now in a working state. Let us know if you need any more changes.

Tedko commented:
@septs @Jack-Works help @Bobface  to rebase?

Bobface commented:
Would be much appreciated!

Jack-Works commented:
the change set is a bit horrifying but let me try to rebase this...

Jack-Works commented:
@MountVase please make sure you have run `pnpm run build` and there is no error. Otherwise, I cannot rebase this PR with confidential I didn't introduce new errors.

guanbinrui committed:
Merge branch 'master' into plugins/ideamarket

guanbinrui committed:
chore: update lock file

guanbinrui committed:
chore: resolve merge errors

guanbinrui commented:
Hi @MountVase, I have resolved those conflicts in #2984. Please merge it or pick those new commits onto your branch.

pbteja1998 committed:
style: change the popover ui

pbteja1998 committed:
feat: show price instead of rank when not hovered

pbteja1998 committed:
chore: remove inline css

pbteja1998 commented:
Hi, please let me know if you need any more changes.
cc: @Bobface

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/define.tsx:
```diff
@@ -0,0 +1,35 @@
+
+        return (
+            <div className={classes.layout}>
+                <Fetcher username={username} />
```
It seems like this plugin will make a request whatever. I'm worrying if it will be the next #2840

pbteja1998 replied:
Yes, you are right. It does make a request for every tweet visible on the viewport. This is because we want to show the Price of the token in line with the tweet without any user interaction. 

@bobface can we do something about this?

Bobface replied:
I don't think so. Our API allows to query multiple listings in a single query, but I don't think there is an easy way to get all Tweets on the screen to make a bundle query.

Jack-Works replied:
Yeah, our API is generally not optimized for this bundle query case. I think we need a user action to check out the price, or the API query must be very fast and cheap otherwise the API server will down cause too many requests

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/Listed.tsx:
```diff
@@ -0,0 +1,85 @@
+                display="flex"
+                justifyContent="space-between"
+                alignItems="center"
+                py={1.375}
```
please do not use this kind of abbr thanks. Use the full term `padding...`

pbteja1998 replied:
Changed it

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/NotListed.tsx:
```diff
@@ -0,0 +1,69 @@
+        zIndex: 1,
+    },
+    topBox: {
+        backgroundColor: '#f7f7f7',
```
Does this UI works in both light and dark mode?

pbteja1998 replied:
Yes, it does.

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/NotListed.tsx:
```diff
@@ -0,0 +1,69 @@
+
+    function clicked(e: MouseEvent) {
+        e.preventDefault()
+        window.open('https://app.ideamarket.io/', '_blank', 'noopener')
```
Please use the `href` instead of this

pbteja1998 replied:
Changed it.

pbteja1998 committed:
fix: use paddingX and paddingY instead of px and py

pbteja1998 committed:
style: add hover color for button

pbteja1998 committed:
fix: use href instead of window.open for ideamarket links

Bobface commented:
Hey guys, checking in on the progress with this PR. Is there something left we need to do?

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/Listed.tsx:
```diff
@@ -31,9 +34,8 @@ interface ListedProps {
-        window.open(`https://ideamarket.io/i/twitter/${props.username}`, '_blank', 'noopener')
+    function clicked(e: SyntheticEvent) {
+        e.stopPropagation()
     }
```
Actually, this part is also useless now.

pbteja1998 replied:
The tweet is also getting clicked if I don't add that.

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/NotListed.tsx:
```diff
@@ -45,9 +46,8 @@ const useStyles = makeStyles((theme) => ({
-        window.open('https://app.ideamarket.io/', '_blank', 'noopener')
+    function clicked(e: SyntheticEvent) {
+        e.stopPropagation()
     }
```
Actually, this part is also useless now.

pbteja1998 replied:
Same here. The tweet is getting clicked when I click on the button without this.

Jack-Works replied:
Oh interesting. So please leave a comment so future devs know it is useful code.

Jack-Works commented on packages/maskbook/src/plugins/Ideamarket/UI/LogoButton.tsx:
```diff
@@ -0,0 +1,110 @@
+    const classes = useStyles()
+
+    const doWeRender = () => {
+        if (clicked) {
+            return (
+                <Listing
+                    username={props.username}
+                    setExtendedHover={setExtendedHover}
+                    found={props.found}
+                    rank={props.rank}
+                    dayChange={props.dayChange}
+                    price={props.price}
+                />
+            )
+        } else if (hover) {
+            return (
+                <Listing
+                    username={props.username}
+                    setExtendedHover={setExtendedHover}
+                    found={props.found}
+                    rank={props.rank}
+                    dayChange={props.dayChange}
+                    price={props.price}
+                />
+            )
+        } else if (extendedHover) {
+            return (
+                <Listing
+                    username={props.username}
+                    setExtendedHover={setExtendedHover}
+                    found={props.found}
+                    rank={props.rank}
+                    dayChange={props.dayChange}
+                    price={props.price}
+                />
+            )
+        } else {
+            return null
+        }
```
```suggestion
        if (clicked || hover || extendedHover) {
            return (
                <Listing
                    username={props.username}
                    setExtendedHover={setExtendedHover}
                    found={props.found}
                    rank={props.rank}
                    dayChange={props.dayChange}
                    price={props.price}
                />
            )
        } else {
            return null
        }
```

Jack-Works replied:
do not repeat yourself

pbteja1998 committed:
chore: remove redundant code

Co-authored-by: Jack Works <jackworks@protonmail.com>

pbteja1998 commented on packages/maskbook/src/plugins/Ideamarket/UI/Listed.tsx:
```diff
@@ -0,0 +1,92 @@
+export default function Listed(props: ListedProps) {
+    const classes = useStyles()
+    function clicked(e: SyntheticEvent) {
+        e.stopPropagation()
```
```suggestion
        e.stopPropagation() // This is to prevent event bubbling. Without this, the tweet will also get clicked. 
```

pbteja1998 committed:
chore: add some useful comments

pbteja1998 commented on packages/maskbook/src/plugins/Ideamarket/UI/NotListed.tsx:
```diff
@@ -0,0 +1,74 @@
+    const classes = useStyles()
+
+    function clicked(e: SyntheticEvent) {
+        e.stopPropagation()
```
```suggestion
        e.stopPropagation() // This is to prevent event bubbling. Without this, the tweet will also get clicked. 
```

pbteja1998 committed:
chore: add comment

Jack-Works commented:
So the code LTGM but I think we should seriously figure the problem of making too many API requests.

Bobface commented:
@Jack-Works What do you think of this approach, where we batch requests and only make one request every second: https://github.com/Ideamarket/Maskbook/commit/d70d6ff33c66dfe84ab501c55d9fda61007ac5aa

Jack-Works commented:
> @Jack-Works What do you think of this approach, where we batch requests and only make one request every second: [Ideamarket@d70d6ff](https://github.com/Ideamarket/Maskbook/commit/d70d6ff33c66dfe84ab501c55d9fda61007ac5aa). Will clean up the code a bit more before I push to this PR, just interested if this would be a possible solution for you.

I don't know if it's really working because it is still a big amount of requests. After all, many users are using Mask. I'm still worrying if the server can take this. How do you think? @yisiliu @guanbinrui @zhouhanseng @septs

yisiliu commented:
> > @Jack-Works What do you think of this approach, where we batch requests and only make one request every second: [Ideamarket@d70d6ff](https://github.com/Ideamarket/Maskbook/commit/d70d6ff33c66dfe84ab501c55d9fda61007ac5aa). Will clean up the code a bit more before I push to this PR, just interested if this would be a possible solution for you.
> 
> I don't know if it's really working because it is still a big amount of requests. After all, many users are using Mask. I'm still worrying if the server can take this. How do you think? @yisiliu @guanbinrui @zhouhanseng @septs

I agree with @Jack-Works. It's alright to make this many requests from our client side but it'll add up to jam the API server. What about adding a cache layer on top of AWS lambda and maintain the "latest" results in a timely(update every 1 or 2 minutes) but also be updated on demand. We don't really want to push too much pressure on the server especially when our user base scales. Now it's around 30-40K weekly active users but we are worried when it goes up to 10x.

Bobface commented:
Okay, we can add a cache between users and our API servers. I will setup something in AWS for that. Also the API is currently loadbalanced across three strong servers, and we can easily add more.

yisiliu commented:
> Okay, we can add a cache between users and our API servers. I will setup something in AWS for that. Also the API is currently loadbalanced across three strong servers, and we can easily add more.

Or let's keep the settings as it is now and add the cache later.

Bobface commented:
I think adding a cache is actually a good idea, I'm setting something up now. Will push a commit which changes the API URL when it's ready and we can try it.

Bobface committed:
feat: use cached api

Bobface committed:
chore: cleanup

Bobface commented:
API is cached on the server side now. I had to swap out `graphql-request` for `@apollo/client` because the former does not support `GET` requests, which is required for AWS API Gateway to be able to cache.

Bobface committed:
Merge branch 'master' into plugins/ideamarket

yisiliu commented:
@Bobface thanks! I think this'll be useful in other applications too. We will try to merge this pr asap in our next release.

pbteja1998 committed:
style: change the style of the Listed component

pbteja1998 committed:
style: remove inter font from Listed overlay

Bobface committed:
Merge pull request #4 from MountVase/more-ui-changes

style: change the style of the Listed component

Jack-Works commented:
hi @MountVase please rebase/merge with master. and @AinuX can u test this after the code up-to-date?

AinuX commented:
@MountVase

pbteja1998 commented:
> @MountVase
> 
> Please update the code, after that I can help to test it.

@AinuX I will be doing it.

pbteja1998 committed:
Merge branch 'master' into plugins/ideamarket

pbteja1998 commented:
@AinuX fixed the conflicts.

Jack-Works reviewed (approved):
@AinuX hi please test this

AinuX commented:
The function is OK, but the UI of popup is different from the original design at https://github.com/Ideamarket/gitcoin-bounties/issues/19.
![截屏2021-05-20 18 03 03](https://user-images.githubusercontent.com/2582974/118961107-8a369d00-b996-11eb-9e79-8c4996bcf669.png)
![截屏2021-05-20 18 03 48](https://user-images.githubusercontent.com/2582974/118961132-8f93e780-b996-11eb-9e05-a5b141705eb3.png)

Bobface commented:
@AinuX That is intentional, we changed the design a bit later on.

guanbinrui commented on packages/maskbook/src/plugins/Ideamarket/UI/Listed.tsx:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    return (
+        <>
+            <div
+                className={classes.root}
+                onMouseEnter={() => props.setExtendedHover(true)}
+                onMouseLeave={() =>
+                    setTimeout(() => {
+                        props.setExtendedHover(false)
+                    }, 200)
+                }>
+                <div className={classes.inner}>
+                    <div className={classes.price}>
+                        <Typography className={classes.priceText}>${props.price}</Typography>
+                    </div>
+                    <Button
+                        onClick={clicked}
+                        href={`https://app.ideamarket.io/i/twitter/${props.username}`}
+                        target="_blank"
+                        rel="noopener"
+                        className={classes.button}>
+                        <Typography className={classes.buttonText}>Buy</Typography>
+                    </Button>
+                </div>
+            </div>
+        </>
```
```suggestion
          <div
              className={classes.root}
              onMouseEnter={() => props.setExtendedHover(true)}
              onMouseLeave={() =>
                  setTimeout(() => {
                      props.setExtendedHover(false)
                  }, 200)
              }>
              <div className={classes.inner}>
                  <div className={classes.price}>
                      <Typography className={classes.priceText}>${props.price}</Typography>
                  </div>
                  <Button
                      onClick={clicked}
                      href={`https://app.ideamarket.io/i/twitter/${props.username}`}
                      target="_blank"
                      rel="noopener"
                      className={classes.button}>
                      <Typography className={classes.buttonText}>Buy</Typography>
                  </Button>
              </div>
          </div>
```

guanbinrui commented on packages/maskbook/src/plugins/Ideamarket/UI/LogoButton.tsx:
```diff
@@ -0,0 +1,88 @@
+        <div className={classes.root}>
+            <IconButton
+                onMouseEnter={() =>
+                    setTimeout(() => {
```
Need clean the previous timer.

guanbinrui commented on packages/maskbook/src/plugins/PluginUI.ts:
```diff
@@ -37,5 +37,6 @@ sideEffect.then(() => {
     if (Flags.airdrop_enabled) plugins.add(AirdropPluginDefine)
     if (Flags.snapshot_enabled) plugins.add(SnapShotPluginDefine)
     if (Flags.collectibles_enabled) plugins.add(CollectiblesPluginDefine)
+    if (Flags.ideamarket_enabled) plugins.add(IdeamarketPluginDefine)
```
Would you please use the new plugin infra?

guanbinrui commented:
Replaced by https://github.com/DimensionDev/Maskbook/pull/5924.
